File: C++/Queue/queue_with_linkedList.cpp
#include<iostream>

using namespace std;

class Node{

	public:
		int data;
		Node* next;
		
		//constructor
		Node(int d){
			data = d;
			next = NULL;
		}
};

class Queue{
	
	private:
		Node* front;
		Node* rear;
	
	public:
		//constructor
		Queue(){
			front = NULL;
			rear = NULL;
		}
		
		bool isEmpty(){
			return front == NULL;
		}
		
		void EnQueue(int data){
			Node* newNode = new Node(data);
			
			if(rear == NULL){
				front = newNode;
				rear = newNode;
			}
			else{
				rear->next = newNode;
				rear = newNode;
			}
		}
		
		int DeQueue(){
			if(isEmpty()){
				cout<<"Queue is empty"<<endl;
				return -1;
			}
			
			Node* temp = front;
			int data = front->data;
			
			if(front == rear){
				front = NULL;
				rear = NULL;
			}
			else{
				front = front->next;
			}
			
			delete temp;
			return data;
		}	
		
		void Display(){
			Node* current = front;
			if (isEmpty()) {
				cout << "Queue is empty." << endl;
				return;
			}
			
			cout<<"Queue: ";
			while(current != NULL){
				cout<<current->data<<" ";
				current = current->next;
			}
			cout<<endl;
		}	
			
};


int main() {
    Queue q;

    q.EnQueue(10);
    q.EnQueue(20);
    q.EnQueue(30);
    q.Display();

    cout << "Dequeued: " << q.DeQueue() << endl;
    q.Display();

    q.EnQueue(40);
    q.Display();

    cout << "Dequeued: " << q.DeQueue() << endl;
    q.Display();

    cout << "Dequeued: " << q.DeQueue() << endl;
    q.Display();

    cout << "Dequeued: " << q.DeQueue() << endl;
    q.Display();

    return 0;
}

File: Python/Queue/queueWithLinkedList.py
class Node:
    def __init__(self, val):
        self.data = val
        self.next = None

class Queue:
        def __init__(self):
            self.front = None
            self.rear = None
            self.size = 0

        def isEmpty(self):
            return self.front is None

        def enQueue(self, val):
            newNode = Node(val)
            if self.rear is None:
                self.front = newNode
                self.rear = newNode
            else:
                self.rear.next = newNode
                self.rear = newNode

            self.size += 1

        def deQueue(self):
            if self.front is None:
                return None
            deququeNode = self.front
            self.front = self.front.next
            if self.front is None:
                self.rear = None
            self.size -= 1
            return deququeNode.data

        def Peek(self):
            if self.front is None:
                return None
            return self.front.data
        
        def __len__(self):
            return self.size

# create a new queue
queue = Queue()

# enqueue some items
queue.enQueue(1)
queue.enQueue(2)
queue.enQueue(3)

# get the size of the queue
print("Queue size: ", len(queue))

# get the front item of the queue
print("Front item: ", queue.Peek())

# dequeue an item from the queue
dequeued_item = queue.deQueue()
print("Dequeued item: ", dequeued_item)

# check if the queue is empty
print("Is queue empty? ", queue.isEmpty())
